File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int target = 0;
        set<vector<int>>s;
        int n = nums.size();
        sort(nums.begin(), nums.end());
        for(int i = 0;i<n-2;i++)
        {
            int j = i + 1;
            int k = nums.size() - 1;
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum == target) {
                    s.insert({nums[i], nums[j], nums[k]});
                    j++;
                    k--;
                } else if (sum < target) {
                    j++;
                } else {
                    k--;
                }
            }
        }
        vector<vector<int>>ans;
        for(auto i : s)
        {
            ans.push_back(i);
        }
        return ans;
    }
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res = []
        nums.sort()
        for i in range(n-2):
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i-1]:
                continue
            l = i+1
            r = len(nums) - 1
            while l < r:
                total = nums[i] + nums[l] + nums[r]
                if total < 0:
                    l += 1
                elif total > 0:
                    r -= 1
                else:
                    temp = [nums[i], nums[l], nums[r]]
                    res.append(temp)
                    while l < r and nums[l] == temp[1]:
                        l += 1
                    while l < r and nums[r] == temp[2]:
                        r-=1
        return res
           